backend/pkg/archiver/archiver.go

```diff
@@ -0,0 +1,126 @@
+			// Don't archive or delete anything for admins
+			continue
+		}
+
+		var archivedDashboards []uint64
+		var activeDashboards []uint64
+
+		type ArchivedDashboard struct {
+			DashboardId    uint64
+			ArchivedReason enums.VDBArchivedReason
+		}
+		var dashboardsToBeArchived []ArchivedDashboard
+		var dashboardsToBeDeleted []uint64
+
+		// Check if the active user dashboard exceeds the maximum number of groups, or validators
+		for _, dashboardInfo := range dashboards {
+			if dashboardInfo.IsArchived {
+				archivedDashboards = append(archivedDashboards, dashboardInfo.DashboardId)
+			} else {
+				if dashboardInfo.GroupCount > userInfo.PremiumPerks.ValidatorGroupsPerDashboard {
+					dashboardsToBeArchived = append(dashboardsToBeArchived, ArchivedDashboard{DashboardId: dashboardInfo.DashboardId, ArchivedReason: enums.VDBArchivedReasons.Groups})
+				} else if dashboardInfo.ValidatorCount > userInfo.PremiumPerks.ValidatorsPerDashboard {
+					dashboardsToBeArchived = append(dashboardsToBeArchived, ArchivedDashboard{DashboardId: dashboardInfo.DashboardId, ArchivedReason: enums.VDBArchivedReasons.Validators})
+				} else {
+					activeDashboards = append(activeDashboards, dashboardInfo.DashboardId)
+				}
+			}
+		}
+
+		// Check if the user still exceeds the maximum number of active dashboards
+		dashboardLimit := int(userInfo.PremiumPerks.ValidatorDashboards)
+		if len(activeDashboards) > dashboardLimit {
+			slices.Sort(activeDashboards)
+			for id := 0; id < len(activeDashboards)-dashboardLimit; id++ {
+				dashboardsToBeArchived = append(dashboardsToBeArchived, ArchivedDashboard{DashboardId: activeDashboards[id], ArchivedReason: enums.VDBArchivedReasons.Dashboards})
+			}
+		}
+
+		// Check if the user exceeds the maximum number of archived dashboards
+		archivedLimit := handlers.MaxArchivedDashboardsCount
+		if len(archivedDashboards)+len(dashboardsToBeArchived) > archivedLimit {
+			dashboardsToBeDeleted = archivedDashboards
+			for _, dashboard := range dashboardsToBeArchived {
+				dashboardsToBeDeleted = append(dashboardsToBeDeleted, dashboard.DashboardId)
+			}
+			slices.Sort(dashboardsToBeDeleted)
+			dashboardsToBeDeleted = dashboardsToBeDeleted[:len(dashboardsToBeDeleted)-archivedLimit]
+		}
+
+		// Archive dashboards
+		dashboardsToBeDeletedMap := utils.SliceToMap(dashboardsToBeDeleted)
+		for _, dashboard := range dashboardsToBeArchived {
+			if _, ok := dashboardsToBeDeletedMap[dashboard.DashboardId]; ok {
+				// The dashboard will immediately be deleted, so no need to archive it
+				continue
+			}
+			archivedReason := dashboard.ArchivedReason
+			_, err := a.das.UpdateValidatorDashboardArchiving(ctx, t.VDBIdPrimary(dashboard.DashboardId), &archivedReason)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Delete dashboards
+		for _, dashboardId := range dashboardsToBeDeleted {
+			err := a.das.RemoveValidatorDashboard(ctx, t.VDBIdPrimary(dashboardId))
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Created `UpdateValidatorDashboardsArchiving` and `RemoveValidatorDashboards`for a single execution but I left `GetUserInfo`with a TODO since it is quite complicated to get the info for all users at once.